zooniverse/front-end-monorepo
# Classifier: Custom Video Controls

~~Blocked because this PR is based on #3645.~~

## Package
lib-classifier

## Linked Issue and/or Talk Post
#3612 and #3613 

## Describe your changes

This is a complete rewrite of the VideoController component. The controls are designed to sit below a html video component so that future drawing tools can be positioned over a video subject. Not all video subject projects will use drawing tools, but for consistency these custom video controls will always display.

### File Re-organization

I consolidated all aspects of the controls into one component: VideoController. It lives in a /components folder in SingleVideoViewer. I also moved the `react-player` component into SingleVideoViewerContainer for readability. All of its handler functions are defined in SingleVideoViewerContainer.

### Controls Functionality

I improved the video scrubbing by memoizing the `react-player` so it's `onProgress` and `setTimeStamp` functions don't call at the same time.

I edited `formatTimeStamp()` to display timestamps in the same units as popular video players such as native html video controls, youtube, vimeo, etc. (i.e. 0:20 means 20 seconds).

I added the ability to change volume and display a video subject in fullscreen.

### Controls Styling

Screenshots for review: 
Note that there is a 1px white border on top of the video controls because the Woodpecker Cavity Cam subjects have black borders, and I want the video controls' dimensions to be obvious. The white border only appears in the screenshots, not in the code.

![woodpecker desktop](https://user-images.githubusercontent.com/23665803/191377047-7ba2c01c-c792-43c7-b1e7-f4161e02b818.jpg)

![woodpecker smaller screen](https://user-images.githubusercontent.com/23665803/191377197-ef5ca0d5-c493-4aab-8df3-94744d12ec5a.jpg)

![woodpecker volume](https://user-images.githubusercontent.com/23665803/191377205-d61031c6-7e2c-441e-a2c3-fec75428708d.jpg)

Should the controls be re-arranged on a mobile view?
![woodpecker mobile](https://user-images.githubusercontent.com/23665803/191377240-8254d51a-4344-487e-abf8-dfa2d800631c.jpg)

### Documentation and Tests

I added documentation in a README file for VideoController and SingleVideoViewerContainer. I added unit tests using React Testing Library for VideoController.

## How to Review
 This PR is part of an effort to improve the video subject viewer code for migrating [Woodpecker Cavity Cam](https://www.zooniverse.org/projects/elwest/woodpecker-cavity-cam) to FEM, so I recommend that project for reviewing functionality.

Things to check a volunteer can do:
- Play / Pause the video
- See the timestamp and duration of video subject
- Scrub the video with the slider
- Change the playback rate
- Change the volume
- View the video subject in fullscreen

Video Controller also appears in SingleVideoViewer's storybook.

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)
